doc/specs/#605 - Search/spec.md

```diff
@@ -0,0 +1,109 @@
+
+## Solution Design
+
+Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:
+
+1. The search is triggered by KeyBindings. A new setting property named "openFind" will be enabled in the Json file. The user can set their own key bindings for search. The default is "ctrl + F". 
+2. The user search in a XAML TextBox, which is contained in a custom SearchBoxControl. The default position of the search box is the top right corner, the user can also move it to bottom right corner to avoid blocking text. 
+3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 
+4. If currently there is no active selection, the search starts from the last line of the mutableViewport. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 
+5. The search dialog should not block terminal's view. In phrase one, this is achieved by make search box movable. The user can move the search box to the bottom right corner.
+6. The user should be able to fully interact with the terminal when the search box is on screen. 
+7. For accessibility concerns, the user should be able to navigate all the iteractive elements on the search box using keyboard tab if the search box is focused. Searchbox could be created and closed with keyboard bindings. Close is usually binding to Esc. 
+8. To avoid searchBox from blocking the terminal window, the search box could be moved to the bottom right by clicking a button. 
+
+Conhost already has a module for search. It realizes case sensitive or insensitive exact text match search, and it provides methods to select the found word. However, we want to make search as a shared component between Terminal and Console host. Now search module is part of Conhost, and its dependencies include BufferOut and some other types in ConHost such as SCREEN_INFORMATION. In order to make Search a shared component, we need to remove its dependency on ConHost types. BufferOut is already a shared component, but we need to make sure there is no other Conhost dependency.
+
+We will create a SearchBoxControl Xaml UserControl element. When a search process begins, a SearchBoxControl object will be created and attach to TermControl root grid. In other words, one SearchBox is added for each TermControl. The reasons for this design is:
+
+1. Each TermControl object is a Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 

```

Comment:
From a lost comment:
> One of the big concerns I still have with this spec is regarding per-terminal search vs global search.
> 
> > One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and **search from all tabs**
> 
> (emph. added)
> 
> With the current design of a per-terminal find dialog, how do we transition to a global (search all terminals) model? I know that global search is out of scope for this first iteration, but if the find dialog is inherently tied to the terminal instance it's searching, how do we then re-use this dialog for searching many terminal instances?
>

Reply:
I discussed with Dustin about in which directory should implement the SearchBoxControl, and we decided to add this control in TerminalControl/ dir. In this way, developers could use this UserControl as an individual component. I think if we wants to implement multi-tab search, we can attach a SearchBoxControl object to the App or TerminalPage, we can provide the text buffer of the current focused terminal. But we need to change the search algorithm.

Reply:
Could we add this to the "Future Considerations" section on search from all tabs? I think that would help answer questions. Just add something about how we could later add a modified `SearchBoxControl` to `TerminalPage` directly to enable a "search from all tabs" option in the future

Reply:
Added.